Interpreter: de-duplicate code

tautschnig opened the pull request:
evaluate_address(byte_extract) and evaluate(byte_extract) had almost the same
implementation.

This commit previously was part of #2165, but I suspect it is somewhat broken as #2165 previously broke test-gen. Awaiting feedback from the automatic reviewer.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 54ed5888e2d9d54b5463baf4e749842946672cc6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85123929

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 9bae4adfc955f0ce240bfc40e0c5c18e2f4ef7a7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86270759

tautschnig commented:
We do not have any tests of the interpreter in the code base (which is a shame), and AFAIK this changed caused some trouble. Assigning to @peterschrammel with the hope to receive some guidance, but if anyone else could contribute tests of the interpreter that would be much appreciated.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f36c08cb6a56c48a6ee20f6fb40d8968c21996da).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102126796
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

thk123 commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -885,42 +885,11 @@ void interpretert::evaluate(
-  else if(expr.id()==ID_byte_extract_little_endian ||
-          expr.id()==ID_byte_extract_big_endian)
-  {
-    if(expr.operands().size()!=2)
```
Just noting this sanity check on byte_extract is being deleted.

thk123 commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -885,42 +885,11 @@ void interpretert::evaluate(
-      mp_integer memory_offset;
-      // If memory offset is found (which should normally be the case)
-      // extract the corresponding data from the appropriate memory location
-      if(!byte_offset_to_memory_offset(
```
You remove this check. No idea what it does, but it should probably be removed in a separate commit to make the move clearer.

thk123 commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -959,16 +928,43 @@ void interpretert::evaluate(
+              numeric_cast_v<std::size_t>(memory_offset + target_type_leaves));
+          return;
+        }
+        // we fail
```
This comment doesn't seem quite right - we don't fail - we carry on to the next if - perhaps a throw here, or a return?

thk123 commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -959,16 +928,43 @@ void interpretert::evaluate(
       {
         dest.resize(numeric_cast_v<std::size_t>(get_size(expr.type())));
         read(address, dest);
+        return;
```
This return is redundant. Probably an indication that just pulling out this whole branch into a function might make it a little easier to follow.

thk123 reviewed (commented):
In principle looks fine (And a preliminary run of TG looks OK too) - but in the interest of helping anyone who ends up git blaming an obsecure bug this introduces could the removal of calling `byte_offset_to_memory_offset` be some how factored into a separate commit.

martin-cs reviewed (approved):
Am fine with the changes but ... can we have a test case because the interpreter is dangerously under tested.

tautschnig committed:
Add regression tests for goto-program interpreter

This first commit just tests that help output can be generated.

tautschnig committed:
goto-program interpreter: accurate help output

Previous help output was misleading and incomplete.

tautschnig committed:
goto-program interpreter: test and fix 'r' command

Fix basic execution as sparse vectors do not (yet) support resizing to
smaller sizes (performing a clear+resize instead).

tautschnig committed:
goto-program interpreter: output memory map at lower verbosity

The memory map is (only) printed upon a user's request (using the "m"
command). It should not be necessary to raise verbosity to debug level
to see this output.

tautschnig committed:
goto-program interpreter: test and fix stepping commands

"s" without a suffix or with a suffix that is not one of [aeo0-9]
previously resulted in an invariant violation. Also make "r0" actually
work by fixing the assumptions about the structure of the top-level
function.

tautschnig committed:
Interpreter: de-duplicate code and make it not fail for unions

evaluate_address(byte_extract) and evaluate(byte_extract) had almost the
same implementation. Instead of the duplication, use `evaluate_address`
to implement `evaluate` and just handle the remaining bits in
`evaluate`.

byte_extract still has a number of problems, as the regression test
shows: addressing works at the object level, not at the level of
individual bytes. Thus we can compute the correct address, but still
wouldn't access the right bytes.

tautschnig committed:
Interpreter: remove unsupported target_assertion

The iterator was never initialised, but still being compared to, which
is undefined behaviour.
